crates/wasmtime/src/store/data.rs

```diff
@@ -191,31 +183,71 @@ where
     }
 }
 
+/// A unique identifier to get attached to a stores.
+///
+/// This identifier is embedded into the `Stored<T>` structure and is used to
+/// identify the original store that items come from. For example a `Memory` is
+/// owned by a `Store` and will embed a `StoreId` internally to say which store
+/// it came from. Comparisons with this value are how panics are generated for
+/// mismatching the item that a store belongs to.
+#[derive(Debug, Copy, Clone, PartialEq)]
+pub struct StoreId(NonZeroU64);
+
+impl StoreId {
+    /// Allocates a new unique identifier for a store that has never before been
+    /// used in this process.
+    fn allocate() -> StoreId {
+        static NEXT_ID: AtomicU64 = AtomicU64::new(0);
+
+        // Only allow 2^63 stores at which point we start panicking to prevent
+        // overflow.
+        //
+        // If a store is created once per microsecond then this will last the
+        // current process for 584,540 years before overflowing.
+        let id = NEXT_ID.fetch_add(1, SeqCst);
```

Comment:
I think this can be a `Relaxed` fetch_add, because the sequence in which IDs are allocated doesn't matter (every ID is as good as any other and is opaque), so there's no need to synchronize this allocation step with respect to any other loads/stores. I wrote a similar justification over [here](https://github.com/bytecodealliance/wasmtime/blob/2a4851ad2b5d77fce41597776157270187fae258/crates/runtime/src/module_id.rs#L28-L39) for `CompiledModuleId`, I guess we could just point to that as justification.